### DISCUSSION TITLE
The correct way to use after_save callback with self-defined error message.

### DISCUSSION THREAD

**Discussion by zw963** [luckyframework/lucky#1886] [Q&A] [answered]
Following is a code snippet.

```cr
after_save code_and_batch_level_must_uniq

  def code_and_batch_level_must_uniq(saved_record : University)
    code.value.try do |code_value|
      batch_level.value.try do |batch_level_value|
        if UniversityQuery.new.code(code_value).batch_level(batch_level_value).select_count > 1
          self.code.add_error("编码和批次的组合, 必须唯一")
          raise Avram::InvalidOperationError.new(self)
        end
      end
    end
  end
```

What I need to ensure is that in the database, same `code` and `batch_level` must uniq.(we assume not add uniq index in db)

Above code work, but will raise 500 if user insert new same code and batch_level record.

So, how to get error message(as add_error in before_save) only instead raise 500 when use input error?

Following situation i have tried:

1. if not raise in after_save, record always create sucessful, not work.
2. if use before_save, like following, we have to determinate current operation if is a new_record, and logic is complicated.

```cr
  before_save code_and_batch_level_must_uniq

  def code_and_batch_level_must_uniq
    code.value.try do |code_value|
      batch_level.value.try do |batch_level_value|
        query = UniversityQuery.new.code(code_value).batch_level(batch_level_value)

        if query.select_count >= 1
          if new_record?
            self.code.add_error("编码和批次的组合, 必须唯一")
          else
            if query.first.id != id.value
              self.code.add_error("编码和批次的组合, 必须唯一")
            end
          end
        end
      end
    end
```

So, i perfer use before_save to achieve this goal.

Thanks

**Comment by jwoertink** [comment, marked as answer]
I'm not completely clear on what the issue is, but when you use `after_save` you are still inside of the initial database transaction. So you can call `database.rollback` which will un-do the original object save.

```crystal
after_save code_and_batch_level_must_uniq

def code_and_batch_level_must_uniq(saved_record : University)
  code.value.try do |code_value|
    if not_unique_code
       code.add_error("must be unique")
       database.rollback  # this line will un-save the university record
     end
  end
end
```

**Reply by zw963** [reply]
Thanks a lot, `database.rollback` is my expected answer.

Do you consider add this into lucky doc? there only a example like following: (not help for my case)

```crystal
class SaveComment < Comment::SaveOperation
  after_save also_update_post

  # If this fails, we roll back the comment save too
  def also_update_post(saved_comment : Comment)
    SavePost.update!(latest_comment: saved_comment.body)
  end
end
```
